src/main/java/org/stellar/sdk/Sep10Challenge.java

```diff
@@ -40,4 +46,366 @@ public static String newChallenge(
+   * @throws InvalidSep10ChallengeException If the SEP-0010 validation fails, the exception will be thrown.
+   * @throws IOException                    If read XDR string fails, the exception will be thrown.
+   */
+  public static List<String> verifyChallengeTransactionThreshold(String challengeTransaction, String serverAccountId, Network network, int threshold, List<Signer> signers) throws InvalidSep10ChallengeException, IOException {
+    String ed25519PublicKeySignerType = "ed25519_public_key";
+
+    if (signers == null || signers.isEmpty()) {
+      throw new InvalidSep10ChallengeException("No signers provided.");
+    }
+
+    List<String> ed25519PublicKeySigners = new ArrayList<String>();
+    for (Signer signer : signers) {
+      if (ed25519PublicKeySignerType.equals(signer.getType())) {
+        ed25519PublicKeySigners.add(signer.getKey());
+      }
+    }
+
+    List<String> signersFound = verifyChallengeTransactionSigners(challengeTransaction, serverAccountId, network, ed25519PublicKeySigners);
+
+    int weight = 0;
+    for (String signer : signersFound) {
+      for (Signer signerWithWeight : signers) {
+        if (signer.equals(signerWithWeight.getKey()) && ed25519PublicKeySignerType.equals(signerWithWeight.getType())) {
+          weight += signerWithWeight.getWeight();
+          break;
+        }
+      }
+    }
+
+    if (weight < threshold) {
+      throw new InvalidSep10ChallengeException(String.format("Signers with weight %d do not meet threshold %d.", weight, threshold));
+    }
+
+    return signersFound;
+  }
+
+  private static List<String> verifyTransactionSignatures(Transaction transaction, List<String> signers) throws InvalidSep10ChallengeException {
+    List<DecoratedSignature> signatures = transaction.getSignatures();
+    if (signatures.isEmpty()) {
+      throw new InvalidSep10ChallengeException("Transaction has no signatures.");
+    }
+    List<String> signersFound = new ArrayList<String>();
+    for (String signer : signers) {
+      if (verifyTransactionSignedByAccountId(transaction, signer)) {
+        signersFound.add(signer);
+      }
+    }
+    return signersFound;
+  }
+
+  private static boolean verifyTransactionSignedByAccountId(Transaction transaction, String accountId) {
+    KeyPair keyPair = KeyPair.fromAccountId(accountId);
+    byte[] txHash = transaction.hash();
+    for (DecoratedSignature decoratedSignature : transaction.getSignatures()) {
+      if (Arrays.equals(decoratedSignature.getHint().getSignatureHint(), keyPair.getSignatureHint().getSignatureHint()) && keyPair.verify(txHash, decoratedSignature.getSignature().getSignature())) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Used to store the results produced by {@link Sep10Challenge#readChallengeTransaction(String, String, Network)}.
+   */
+  public static class ChallengeTransaction {
+    private Transaction transaction;
```

Comment:
it seems like `ChallengeTransaction` should be immutable. I would make these fields final and remove their setters